### PR TITLE
Inferred spans

### PR DESCRIPTION
**Pull request by JonasKunz**
**Description:**

This PR adds an OpenTelemetry SDK Extension (SpanProcessor) for enriching traces with "inferred" spans derived from profiling data. It internally uses [async-profiler 3.0](https://github.com/async-profiler/async-profiler) in wall-clock profiling mode.

This feature was originally added a few years ago to the [Elastic APM Agent](https://github.com/elastic/apm-agent-java/tree/main/apm-agent-plugins/apm-profiling-plugin), we recently ported it to OpenTelemetry in our OpenTelemetry distro and would now like to contribute it to the OpenTelemetry community.

The feature works by keeping track of which spans are activated/deactivated on which threads via the OpenTelemetry context API. A log of these activations/deactivations is spilled to disk, while at the same time async-profiler is enabled to profile threads with active spans. After the profiling session ends (default is 10 seconds), the profiling JFR and the log of span activations is used to generate the synthetic inferred spans. [Here](https://www.elastic.co/observability-labs/blog/tracing-data-inferred-spans-opentelemetry) is also a blogpost about the feature.

Because this feature was initially written for the classic Elastic APM Agent, it heavily tries to minimize allocations. As a result, the feature is almost allocation free, which unfortunately comes at the cost of a bit more complex code due to pooling.

**Testing:**

The algorithm for span inference has been ported without changes from the classic Elastic APM Agent and therefore is relatively battle tested. The PR also comes with a large set of test-cases for edge cases of the inference algorithm.

**Documentation:**

Just the README.MD added on how to use this extension either by manual setup or using SDK auto configuration.
It also documents some limitations of the inference algorithm.

### CODE REVIEW

File: inferred-spans/src/main/java/io/opentelemetry/contrib/inferredspans/internal/semconv/Attributes.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+public class Attributes {
+
+  private Attributes() {}
+
+  public static final AttributeKey<String> CODE_STACKTRACE =
+      AttributeKey.stringKey("code.stacktrace");
+  public static final AttributeKey<Boolean> LINK_IS_CHILD = AttributeKey.booleanKey("is_child");
+  public static final AttributeKey<Boolean> SPAN_IS_INFERRED =
+      AttributeKey.booleanKey("is_inferred");
```
Comment (JonasKunz) [inline inferred-spans/src/main/java/io/opentelemetry/contrib/inferredspans/internal/semconv/Attributes.java:18]: `is_child` and `is_inferred` are feature-specific attributes which I don't see as candidated for SemConv.
Is there a best-practice for namespacing such attributes? E.g. to use the feature name as prefix (`inferred_spans.`)?

File: inferred-spans/src/main/java/io/opentelemetry/contrib/inferredspans/InferredSpansAutoConfig.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+  static final String DIAGNOSTIC_FILES_OPTION = "otel.inferred.spans.backup.diagnostic.files";
+  static final String SAFEMODE_OPTION = "otel.inferred.spans.safe.mode";
+  static final String POSTPROCESSING_OPTION = "otel.inferred.spans.post.processing.enabled";
+  static final String SAMPLING_INTERVAL_OPTION = "otel.inferred.spans.sampling.interval";
+  static final String MIN_DURATION_OPTION = "otel.inferred.spans.min.duration";
+  static final String INCLUDED_CLASSES_OPTION = "otel.inferred.spans.included.classes";
+  static final String EXCLUDED_CLASSES_OPTION = "otel.inferred.spans.excluded.classes";
+  static final String INTERVAL_OPTION = "otel.inferred.spans.interval";
+  static final String DURATION_OPTION = "otel.inferred.spans.duration";
+  static final String LIB_DIRECTORY_OPTION = "otel.inferred.spans.lib.directory";
```
Comment (JonasKunz) [inline inferred-spans/src/main/java/io/opentelemetry/contrib/inferredspans/InferredSpansAutoConfig.java:36]: Is there any recommended namespace here to use for the config options? Or is `otel.inferred.spans` fine?

**Review by jackshirazi — APPROVED** [review]

**Review by SylvainJuge — APPROVED** [review]

File: inferred-spans/src/test/java/io/opentelemetry/contrib/inferredspans/internal/asyncprofiler/JfrParserTest.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    JfrParser jfrParser = new JfrParser(ByteBuffer.allocate(368), ByteBuffer.allocate(368));
+
+    File file =
+        Paths.get(JfrParserTest.class.getClassLoader().getResource("recording.jfr").toURI())
+            .toFile();
+
+    jfrParser.parse(
+        file,
+        Collections.emptyList(),
+        Collections.singletonList(caseSensitiveMatcher("co.elastic.otel.*")));
```
Comment (trask) [inline inferred-spans/src/test/java/io/opentelemetry/contrib/inferredspans/internal/asyncprofiler/JfrParserTest.java:37]: should this be the new package name?

File: inferred-spans/src/main/java/io/opentelemetry/contrib/inferredspans/internal/pooling/ObjectPool.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+package io.opentelemetry.contrib.inferredspans.internal.pooling;
+
+import org.jctools.queues.MpmcArrayQueue;
+
+/**
+ * Object pool
+ *
+ * @param <T> pooled object type. Does not have to implement {@link Recyclable} in order to allow
+ *     for dealing with objects that are outside of elastic apm agent (like standard JDK or third
```
Comment (trask) [inline inferred-spans/src/main/java/io/opentelemetry/contrib/inferredspans/internal/pooling/ObjectPool.java:14]:
```suggestion
 *     for dealing with external objects (like standard JDK or third
```

File: inferred-spans/README.md [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+If you manually set-up your `OpenTelemetrySDK`, you need to create and register an `InferredSpansProcessor` with your `TracerProvider`:
+
+```java
+InferredSpansProcessor inferredSpans = InferredSpansProcessor.builder()
+  //.samplingInterval(Duration.ofMillis(10))
+  .build();
+SdkTracerProvider tracerProvider = SdkTracerProvider.builder()
+  .addSpanProcessor(inferredSpans)
+  .addSpanProcessor(BatchSpanProcessor.builder(OtlpGrpcSpanExporter.builder()
+    .setEndpoint("https://<clusterid>.apm.europe-west3.gcp.cloud.es.io:443")
```
Comment (trask) [inline inferred-spans/README.md:56]:
```suggestion
    .setEndpoint("http://localhost:4317")
```

**Review by trask — APPROVED** [review]

File: inferred-spans/src/test/java/io/opentelemetry/contrib/inferredspans/internal/SamplingProfilerTest.java [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+    assertThat(getProfilerTempFiles())
+        .describedAs("should delete temp files when profiler is stopped")
+        .isEmpty();
+  }
+
+  private static List<Path> getProfilerTempFiles() {
+    Path tempFolder = Paths.get(System.getProperty("java.io.tmpdir"));
+    try (Stream<Path> files = Files.list(tempFolder)) {
+      return files
+          .filter(f -> f.getFileName().toString().startsWith("apm-"))
```
Comment (trask) [inline inferred-spans/src/test/java/io/opentelemetry/contrib/inferredspans/internal/SamplingProfilerTest.java:101]:
```suggestion
          .filter(f -> f.getFileName().toString().startsWith("otel-inferred-"))
```

File: inferred-spans/src/test/java/io/opentelemetry/contrib/inferredspans/internal/SamplingProfilerTest.java [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+    // unlike the temporary files that are created by profiler itself
+
+    InferredSpansConfiguration defaultConfig;
+    try (InferredSpansProcessor profiler1 =
+        InferredSpansProcessor.builder().startScheduledProfiling(false).build()) {
+      defaultConfig = ProfilerTestSetup.extractProfilerImpl(profiler1).getConfig();
+    }
+
+    Path tempFile1 = Files.createTempFile("apm-provided", "test.bin");
+    Path tempFile2 = Files.createTempFile("apm-provided", "test.jfr");
```
Comment (trask) [inline inferred-spans/src/test/java/io/opentelemetry/contrib/inferredspans/internal/SamplingProfilerTest.java:121]:
```suggestion
    Path tempFile1 = Files.createTempFile("otel-inferred-provided", "test.bin");
    Path tempFile2 = Files.createTempFile("otel-inferred-provided", "test.jfr");
```

File: inferred-spans/src/test/resources/recording.jfr
Comment (trask) [inline inferred-spans/src/test/resources/recording.jfr:1]: can you include instructions on how to generate this file?

**Review by SylvainJuge — APPROVED** [review]

**Comment by trask** [comment]
@JonasKunz sorry, can you resolve new merge conflict? thanks

**Review by SylvainJuge — APPROVED** [review]
